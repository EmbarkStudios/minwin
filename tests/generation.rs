// use tracing_subscriber::{prelude::__tracing_subscriber_SubscriberExt, Layer};
//     let layer = tracing_subscriber::fmt::layer()
//         .with_test_writer()
//         .with_filter(tracing::metadata::LevelFilter::from_level(
//             tracing::metadata::Level::TRACE,
//         ));
//     let reg = tracing_subscriber::Registry::default().with(layer);
//     let dis = tracing::Dispatch::new(reg);
//     //let _guard = tracing::dispatcher::set_global_default(dis);

use anyhow::Context as _;

fn generate(input: &str, format: bool, convert_case: bool) {
    let mut parser = minwin::Parser::default();
    parser.add_file(format!("tests/data/{input}.rs"));
    let mut parsed = parser.parse().pop().unwrap();

    let mut hints = minwin::Hints::default();
    parsed.gather_hints(&mut hints);

    let md = minwin::MetadataFiles::new().expect("failed to gather metadata files");
    let resolved = minwin::Resolver::flatten(&md, hints).expect("failed to resolve metadata");

    let generated = parsed
        .iter_bind_modules()
        .enumerate()
        .map(|(i, m)| {
            let ts = minwin::generate(&resolved, m, convert_case)
                .with_context(|| format!("{}", m.ident))?;
            Ok((i, ts))
        })
        .collect::<anyhow::Result<Vec<_>>>()
        .expect("failed to generate modules");

    for (index, ts) in generated {
        parsed
            .replace_module(index, ts)
            .expect("failed to replace {index}");
    }

    let generated = parsed.generate(format).expect("failed to generate");

    insta::assert_snapshot!(format!("{input}_{format}"), generated);
}

/// Ensures we can generate function pointers for regular functions for use
/// with eg. `GetProcAddress`
#[test]
fn delegates_for_functions() {
    generate("parking_lot", true, false);
}

/// Ensures we can generate more complex types
/// 1. Arch specific structs
/// 2. Records with alignment
/// 3. Records with packing
/// 4. Nested structs/unions
#[test]
fn complex() {
    generate("md_writer", true, false);
}

/// While the windows metadata professes to encourage enums instead of plain
/// integer constants, there...aren't that many. But just check we can emit the
/// basics
#[test]
fn enums() {
    generate("enums", true, false);
}

/// Ensures we can emit function pointers for actual function pointers
#[test]
fn delegates() {
    generate("delegates", true, false);
}
