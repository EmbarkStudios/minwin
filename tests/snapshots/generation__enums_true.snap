---
source: tests/generation.rs
expression: generated
---
#[minwin]
mod enums {
    pub type WIN32_ERROR = u32;
    pub const ERROR_OUTOFMEMORY: WIN32_ERROR = 14;
    pub type DUPLICATE_HANDLE_OPTIONS = u32;
    pub const DUPLICATE_CLOSE_SOURCE: DUPLICATE_HANDLE_OPTIONS = 1;
    pub const DUPLICATE_SAME_ACCESS: DUPLICATE_HANDLE_OPTIONS = 2;
    pub type FILE_NOTIFY_CHANGE = u32;
    pub const FILE_NOTIFY_CHANGE_FILE_NAME: FILE_NOTIFY_CHANGE = 1;
    pub const FILE_NOTIFY_CHANGE_SIZE: FILE_NOTIFY_CHANGE = 8;
    pub const FILE_NOTIFY_CHANGE_LAST_WRITE: FILE_NOTIFY_CHANGE = 16;
    pub type ALG = u32;
    pub const ALG_CLASS_ANY: ALG = 0;
    pub const ALG_CLASS_SIGNATURE: ALG = 8192;
    pub const ALG_CLASS_MSG_ENCRYPT: ALG = 16384;
    pub const ALG_CLASS_DATA_ENCRYPT: ALG = 24576;
    pub const ALG_CLASS_HASH: ALG = 32768;
    pub const ALG_CLASS_KEY_EXCHANGE: ALG = 40960;
    pub const ALG_CLASS_ALL: ALG = 57344;
    pub const ALG_TYPE_ANY: ALG = 0;
    pub const ALG_TYPE_DSS: ALG = 512;
    pub const ALG_TYPE_RSA: ALG = 1024;
    pub const ALG_TYPE_BLOCK: ALG = 1536;
    pub const ALG_TYPE_STREAM: ALG = 2048;
    pub const ALG_TYPE_DH: ALG = 2560;
    pub const ALG_TYPE_SECURECHANNEL: ALG = 3072;
    pub const ALG_TYPE_ECDH: ALG = 3584;
    pub const ALG_TYPE_THIRDPARTY: ALG = 4096;
    pub const ALG_SID_ANY: ALG = 0;
    pub const ALG_SID_THIRDPARTY_ANY: ALG = 0;
    pub const ALG_SID_RSA_ANY: ALG = 0;
    pub const ALG_SID_RSA_PKCS: ALG = 1;
    pub const ALG_SID_RSA_MSATWORK: ALG = 2;
    pub const ALG_SID_RSA_ENTRUST: ALG = 3;
    pub const ALG_SID_RSA_PGP: ALG = 4;
    pub const ALG_SID_DSS_ANY: ALG = 0;
    pub const ALG_SID_DSS_PKCS: ALG = 1;
    pub const ALG_SID_DSS_DMS: ALG = 2;
    pub const ALG_SID_ECDSA: ALG = 3;
    pub const ALG_SID_DES: ALG = 1;
    pub const ALG_SID_3DES: ALG = 3;
    pub const ALG_SID_DESX: ALG = 4;
    pub const ALG_SID_IDEA: ALG = 5;
    pub const ALG_SID_CAST: ALG = 6;
    pub const ALG_SID_SAFERSK64: ALG = 7;
    pub const ALG_SID_SAFERSK128: ALG = 8;
    pub const ALG_SID_3DES_112: ALG = 9;
    pub const ALG_SID_CYLINK_MEK: ALG = 12;
    pub const ALG_SID_RC5: ALG = 13;
    pub const ALG_SID_AES_128: ALG = 14;
    pub const ALG_SID_AES_192: ALG = 15;
    pub const ALG_SID_AES_256: ALG = 16;
    pub const ALG_SID_AES: ALG = 17;
    pub const ALG_SID_SKIPJACK: ALG = 10;
    pub const ALG_SID_TEK: ALG = 11;
    pub const ALG_SID_RC2: ALG = 2;
    pub const ALG_SID_RC4: ALG = 1;
    pub const ALG_SID_SEAL: ALG = 2;
    pub const ALG_SID_DH_SANDF: ALG = 1;
    pub const ALG_SID_DH_EPHEM: ALG = 2;
    pub const ALG_SID_AGREED_KEY_ANY: ALG = 3;
    pub const ALG_SID_KEA: ALG = 4;
    pub const ALG_SID_ECDH: ALG = 5;
    pub const ALG_SID_ECDH_EPHEM: ALG = 6;
    pub const ALG_SID_MD2: ALG = 1;
    pub const ALG_SID_MD4: ALG = 2;
    pub const ALG_SID_MD5: ALG = 3;
    pub const ALG_SID_SHA: ALG = 4;
    pub const ALG_SID_SHA1: ALG = 4;
    pub const ALG_SID_MAC: ALG = 5;
    pub const ALG_SID_RIPEMD: ALG = 6;
    pub const ALG_SID_RIPEMD160: ALG = 7;
    pub const ALG_SID_SSL3SHAMD5: ALG = 8;
    pub const ALG_SID_HMAC: ALG = 9;
    pub const ALG_SID_TLS1PRF: ALG = 10;
    pub const ALG_SID_HASH_REPLACE_OWF: ALG = 11;
    pub const ALG_SID_SHA_256: ALG = 12;
    pub const ALG_SID_SHA_384: ALG = 13;
    pub const ALG_SID_SHA_512: ALG = 14;
    pub const ALG_SID_SSL3_MASTER: ALG = 1;
    pub const ALG_SID_SCHANNEL_MASTER_HASH: ALG = 2;
    pub const ALG_SID_SCHANNEL_MAC_KEY: ALG = 3;
    pub const ALG_SID_PCT1_MASTER: ALG = 4;
    pub const ALG_SID_SSL2_MASTER: ALG = 5;
    pub const ALG_SID_TLS1_MASTER: ALG = 6;
    pub const ALG_SID_SCHANNEL_ENC_KEY: ALG = 7;
    pub const ALG_SID_ECMQV: ALG = 1;
    pub const ALG_SID_EXAMPLE: ALG = 80;
    pub type PROP_ID = u32;
    pub const CERT_KEY_PROV_HANDLE_PROP_ID: PROP_ID = 1;
    pub const CERT_KEY_PROV_INFO_PROP_ID: PROP_ID = 2;
    pub const CERT_SHA1_HASH_PROP_ID: PROP_ID = 3;
    pub const CERT_MD5_HASH_PROP_ID: PROP_ID = 4;
    pub const CERT_HASH_PROP_ID: PROP_ID = 3;
    pub const CERT_KEY_CONTEXT_PROP_ID: PROP_ID = 5;
    pub const CERT_KEY_SPEC_PROP_ID: PROP_ID = 6;
    pub const CERT_IE30_RESERVED_PROP_ID: PROP_ID = 7;
    pub const CERT_PUBKEY_HASH_RESERVED_PROP_ID: PROP_ID = 8;
    pub const CERT_ENHKEY_USAGE_PROP_ID: PROP_ID = 9;
    pub const CERT_CTL_USAGE_PROP_ID: PROP_ID = 9;
    pub const CERT_NEXT_UPDATE_LOCATION_PROP_ID: PROP_ID = 10;
    pub const CERT_FRIENDLY_NAME_PROP_ID: PROP_ID = 11;
    pub const CERT_PVK_FILE_PROP_ID: PROP_ID = 12;
    pub const CERT_DESCRIPTION_PROP_ID: PROP_ID = 13;
    pub const CERT_ACCESS_STATE_PROP_ID: PROP_ID = 14;
    pub const CERT_SIGNATURE_HASH_PROP_ID: PROP_ID = 15;
    pub const CERT_SMART_CARD_DATA_PROP_ID: PROP_ID = 16;
    pub const CERT_EFS_PROP_ID: PROP_ID = 17;
    pub const CERT_FORTEZZA_DATA_PROP_ID: PROP_ID = 18;
    pub const CERT_ARCHIVED_PROP_ID: PROP_ID = 19;
    pub const CERT_KEY_IDENTIFIER_PROP_ID: PROP_ID = 20;
    pub const CERT_AUTO_ENROLL_PROP_ID: PROP_ID = 21;
    pub const CERT_PUBKEY_ALG_PARA_PROP_ID: PROP_ID = 22;
    pub const CERT_CROSS_CERT_DIST_POINTS_PROP_ID: PROP_ID = 23;
    pub const CERT_ISSUER_PUBLIC_KEY_MD5_HASH_PROP_ID: PROP_ID = 24;
    pub const CERT_SUBJECT_PUBLIC_KEY_MD5_HASH_PROP_ID: PROP_ID = 25;
    pub const CERT_ENROLLMENT_PROP_ID: PROP_ID = 26;
    pub const CERT_DATE_STAMP_PROP_ID: PROP_ID = 27;
    pub const CERT_ISSUER_SERIAL_NUMBER_MD5_HASH_PROP_ID: PROP_ID = 28;
    pub const CERT_SUBJECT_NAME_MD5_HASH_PROP_ID: PROP_ID = 29;
    pub const CERT_EXTENDED_ERROR_INFO_PROP_ID: PROP_ID = 30;
    pub const CERT_RENEWAL_PROP_ID: PROP_ID = 64;
    pub const CERT_ARCHIVED_KEY_HASH_PROP_ID: PROP_ID = 65;
    pub const CERT_AUTO_ENROLL_RETRY_PROP_ID: PROP_ID = 66;
    pub const CERT_AIA_URL_RETRIEVED_PROP_ID: PROP_ID = 67;
    pub const CERT_AUTHORITY_INFO_ACCESS_PROP_ID: PROP_ID = 68;
    pub const CERT_BACKED_UP_PROP_ID: PROP_ID = 69;
    pub const CERT_OCSP_RESPONSE_PROP_ID: PROP_ID = 70;
    pub const CERT_REQUEST_ORIGINATOR_PROP_ID: PROP_ID = 71;
    pub const CERT_SOURCE_LOCATION_PROP_ID: PROP_ID = 72;
    pub const CERT_SOURCE_URL_PROP_ID: PROP_ID = 73;
    pub const CERT_NEW_KEY_PROP_ID: PROP_ID = 74;
    pub const CERT_OCSP_CACHE_PREFIX_PROP_ID: PROP_ID = 75;
    pub const CERT_SMART_CARD_ROOT_INFO_PROP_ID: PROP_ID = 76;
    pub const CERT_NO_AUTO_EXPIRE_CHECK_PROP_ID: PROP_ID = 77;
    pub const CERT_NCRYPT_KEY_HANDLE_PROP_ID: PROP_ID = 78;
    pub const CERT_HCRYPTPROV_OR_NCRYPT_KEY_HANDLE_PROP_ID: PROP_ID = 79;
    pub const CERT_SUBJECT_INFO_ACCESS_PROP_ID: PROP_ID = 80;
    pub const CERT_CA_OCSP_AUTHORITY_INFO_ACCESS_PROP_ID: PROP_ID = 81;
    pub const CERT_CA_DISABLE_CRL_PROP_ID: PROP_ID = 82;
    pub const CERT_ROOT_PROGRAM_CERT_POLICIES_PROP_ID: PROP_ID = 83;
    pub const CERT_ROOT_PROGRAM_NAME_CONSTRAINTS_PROP_ID: PROP_ID = 84;
    pub const CERT_SUBJECT_OCSP_AUTHORITY_INFO_ACCESS_PROP_ID: PROP_ID = 85;
    pub const CERT_SUBJECT_DISABLE_CRL_PROP_ID: PROP_ID = 86;
    pub const CERT_CEP_PROP_ID: PROP_ID = 87;
    pub const CERT_SIGN_HASH_CNG_ALG_PROP_ID: PROP_ID = 89;
    pub const CERT_SCARD_PIN_ID_PROP_ID: PROP_ID = 90;
    pub const CERT_SCARD_PIN_INFO_PROP_ID: PROP_ID = 91;
    pub const CERT_SUBJECT_PUB_KEY_BIT_LENGTH_PROP_ID: PROP_ID = 92;
    pub const CERT_PUB_KEY_CNG_ALG_BIT_LENGTH_PROP_ID: PROP_ID = 93;
    pub const CERT_ISSUER_PUB_KEY_BIT_LENGTH_PROP_ID: PROP_ID = 94;
    pub const CERT_ISSUER_CHAIN_SIGN_HASH_CNG_ALG_PROP_ID: PROP_ID = 95;
    pub const CERT_ISSUER_CHAIN_PUB_KEY_CNG_ALG_BIT_LENGTH_PROP_ID: PROP_ID = 96;
    pub const CERT_NO_EXPIRE_NOTIFICATION_PROP_ID: PROP_ID = 97;
    pub const CERT_AUTH_ROOT_SHA256_HASH_PROP_ID: PROP_ID = 98;
    pub const CERT_NCRYPT_KEY_HANDLE_TRANSFER_PROP_ID: PROP_ID = 99;
    pub const CERT_HCRYPTPROV_TRANSFER_PROP_ID: PROP_ID = 100;
    pub const CERT_SMART_CARD_READER_PROP_ID: PROP_ID = 101;
    pub const CERT_SEND_AS_TRUSTED_ISSUER_PROP_ID: PROP_ID = 102;
    pub const CERT_KEY_REPAIR_ATTEMPTED_PROP_ID: PROP_ID = 103;
    pub const CERT_DISALLOWED_FILETIME_PROP_ID: PROP_ID = 104;
    pub const CERT_ROOT_PROGRAM_CHAIN_POLICIES_PROP_ID: PROP_ID = 105;
    pub const CERT_SMART_CARD_READER_NON_REMOVABLE_PROP_ID: PROP_ID = 106;
    pub const CERT_SHA256_HASH_PROP_ID: PROP_ID = 107;
    pub const CERT_SCEP_SERVER_CERTS_PROP_ID: PROP_ID = 108;
    pub const CERT_SCEP_RA_SIGNATURE_CERT_PROP_ID: PROP_ID = 109;
    pub const CERT_SCEP_RA_ENCRYPTION_CERT_PROP_ID: PROP_ID = 110;
    pub const CERT_SCEP_CA_CERT_PROP_ID: PROP_ID = 111;
    pub const CERT_SCEP_SIGNER_CERT_PROP_ID: PROP_ID = 112;
    pub const CERT_SCEP_NONCE_PROP_ID: PROP_ID = 113;
    pub const CERT_SCEP_ENCRYPT_HASH_CNG_ALG_PROP_ID: PROP_ID = 114;
    pub const CERT_SCEP_FLAGS_PROP_ID: PROP_ID = 115;
    pub const CERT_SCEP_GUID_PROP_ID: PROP_ID = 116;
    pub const CERT_SERIALIZABLE_KEY_CONTEXT_PROP_ID: PROP_ID = 117;
    pub const CERT_ISOLATED_KEY_PROP_ID: PROP_ID = 118;
    pub const CERT_SERIAL_CHAIN_PROP_ID: PROP_ID = 119;
    pub const CERT_KEY_CLASSIFICATION_PROP_ID: PROP_ID = 120;
    pub const CERT_OCSP_MUST_STAPLE_PROP_ID: PROP_ID = 121;
    pub const CERT_DISALLOWED_ENHKEY_USAGE_PROP_ID: PROP_ID = 122;
    pub const CERT_NONCOMPLIANT_ROOT_URL_PROP_ID: PROP_ID = 123;
    pub const CERT_PIN_SHA256_HASH_PROP_ID: PROP_ID = 124;
    pub const CERT_CLR_DELETE_KEY_PROP_ID: PROP_ID = 125;
    pub const CERT_NOT_BEFORE_FILETIME_PROP_ID: PROP_ID = 126;
    pub const CERT_NOT_BEFORE_ENHKEY_USAGE_PROP_ID: PROP_ID = 127;
    pub const CERT_DISALLOWED_CA_FILETIME_PROP_ID: PROP_ID = 128;
    pub const CERT_FIRST_RESERVED_PROP_ID: PROP_ID = 129;
    pub const CERT_LAST_RESERVED_PROP_ID: PROP_ID = 32767;
    pub const CERT_FIRST_USER_PROP_ID: PROP_ID = 32768;
    pub const CERT_LAST_USER_PROP_ID: PROP_ID = 65535;
    pub const CERT_STORE_LOCALIZED_NAME_PROP_ID: PROP_ID = 4096;
    #[link(name = "kernel32")]
    extern "system" {
        pub fn GetLastError() -> WIN32_ERROR;
    }
}

